### ISSUE TITLE
Showing method call arguments in ExpectationException message

### ISSUE THREAD

**Issue by lijunle** [FakeItEasy/FakeItEasy#492]
Hi, FakeItEasy owner. Thanks for making such a great Framework.

I have a feature request for strict mode. When it fails on ExpectationException, besides showing the calling method name in the exception message, could you please showing the calling arguments?

That information is so important and much useful for debugging.

**Comment by blairconrad** [comment]
Thanks for the issue @lijunle. Can you expand a bit, perhaps? What scenario are you encountering where just knowing that the method was called is not sufficient?
Do you have an example you can share?

**Comment by lijunle** [comment]
Hi, @blairconrad . Thanks for your reply.

As said in [strict mode wiki page](https://github.com/FakeItEasy/FakeItEasy/wiki/Strict-Fakes):

> Strict fakes are useful when it is important to ensure that no calls are made to your fake other than the ones you are expecting.

I am use strict mode to ensure the call passing right arguments to the mocking method. If it does not, I get the message from ExpectationException and wrap it to Xunit exception to fail the test case. it works well, but the only problem is when the passing arguments are not matching my preset, it only shows the method name, I need to guess what does the arguments is passed and how to fix it.

I know that sometimes it is hard to stringify the method arguments, but at least store it as ExpectationException property. Then when debug, I can set breakpoint and watch the details.

So, that is my story. Hope you add such feature support.

**Comment by blairconrad** [comment]
@lijunle, of course. Excellent example. I tend not to be a `Strict` user, so that hadn't occurred to me.

**Comment by jimmyheaddon** [comment]
@blairconrad - is this in the same ballpark to my other two pull requests, where we've added contextual information to exception messages?  If @lijunle can provide a working example I'd be happy to see if there's a simple solution?

**Comment by blairconrad** [comment]
@jimmyheaddon, I haven't looked deeply into it, but probably similar. It may not be providing an exception message directly, but providing a failure message which FakeItEasy will build into an exception.

I think the site that's not performing how @lijunle would like is `FakeOptionsBuilderExtensions.Strict`.
The `CallWriter` may be a help. May.

**Comment by jimmyheaddon** [comment]
@blairconrad would this be eligible for 2.0 if we can get something in for this?

ExpectationException's are thrown from 3 sites:
```
1. FakeAsserter
2. OrderedFakeAsserter
3. FakeOptionsBuilderExtensions.Strict
```

Are we talking about updating only the latter of the 3 calls (and possibly the spelling/grammatical mistake in ```OrderedFakeAsserter ``` on line 71 - *"The calls* **where** *found but not in the correct order among the calls:"*)?

The use of ```CallWriter``` makes sense, we use it in ```FakeAsserter``` already.

**Comment by blairconrad** [comment]
@jimmyheaddon, I believe that only the strict calls suffer from the problem. When we assert ordered calls and fail, I'm seeing the arguments spewed, using a modified version of the example in [the docs](https://github.com/FakeItEasy/FakeItEasy/wiki/Ordered-assertions). Have you seen different, either for ordered assertions or non-ordered normal assertions?

```
FakeItEasy.ExpectationException : 

  Assertion failed for the following calls:
    'FakeItEasyQuestions.IDoSomethingPrettyUseful.JustDoIt("bah")' repeated at least once
    'FakeItEasyQuestions.IUnitOfWorkFactory.BeginWork()' repeated at least once
  The calls where found but not in the correct order among the calls:
    1: FakeItEasyQuestions.IUnitOfWorkFactory.BeginWork()
    2: FakeItEasyQuestions.IDoSomethingPrettyUseful.JustDoIt(doWhat: "bah")
    3: System.IDisposable.Dispose()
```

I like the idea of fixing the spelling mistake, but I'd be inclined to spin if off into its own work (we wouldn't need an issue; a PR would be fine). Even moreso if we don't have to touch `OrderedFakeAsserter`.
Good catch.

**Comment by blairconrad** [comment]
Oh, I didn't address your question about eligibility for 2.0.
Certainly it would be.  I wouldn't turn down community-supplied improvements. And there's no hard date for 2.0.
That being said, I wouldn't consider it a high priority. Me, I'd prefer to focus on the last few breaking changes that we want to put in 2.0 in (we're following Semantic Versioning, you see, so we want to squeeze as much breakage into 2.0 as possible) and would be okay if some bugfixes and enhancements waited until 2.1 or 2.0.1 or whatever.
I'm not trying to dissuade you from working on anything! It's always nice to get contributions, and you seem keen and also friendly, so that's a plus.
And I may not even be speaking for the other owners regarding scope—this is just my preference. (Personally, I'd secretly been hoping for a March 2015 2.0 release or thereabouts, but we found extra things we wanted to break. Also the outside world intrudes…)

**Comment by adamralph** [comment]
I've fixed the "were" spelling in #778

**Comment by adamralph** [comment]
Should be on 3.2.0?

**Comment by thomaslevesque** [comment]
> Should be on 3.2.0?

👍

**Comment by blairconrad** [comment]
Oh, sorry. I thought I'd done that. I've been trying to keep up with the milestone and release when an issue is closed.

**Comment by blairconrad** [comment]
This issue was fixed in [Release 3.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.2.0).

@lijunle, look for your name in the release notes! :trophy:
